### PR TITLE
Libcloud 911 oneandone compute driver

### PR DESCRIPTION
**Pull request by jasmingacic**
Libcloud 911 oneandone compute driver

### Description

Implementation of  compute driver for [1&1](https://1and1.com/) Provider

### Status

- done, ready for review

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [x] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

### CODE REVIEW

File: libcloud/common/base.py [hunk 1]
```diff
@@ -486,6 +486,7 @@ def user_agent_append(self, token):
 
     def request(self, action, params=None, data=None, headers=None,
                 method='GET', raw=False, stream=False):
+
```
Comment (tonybaloney) [inline libcloud/common/base.py]: looks like you added a blank line before the docstring, travis will probably fail this

File: libcloud/compute/providers.py [hunk 2]
```diff
@@ -143,6 +143,8 @@
     ('libcloud.compute.drivers.ecs', 'ECSDriver'),
     Provider.CLOUDSCALE:
     ('libcloud.compute.drivers.cloudscale', 'CloudscaleNodeDriver'),
+    Provider.ONEANDONE:
+        ('libcloud.compute.drivers.oneandone', 'OneAndOneNodeDriver'),
```
Comment (tonybaloney) [inline libcloud/compute/providers.py]: pls fix alignment

**Review by tonybaloney — CHANGES_REQUESTED** [review]

**Comment by tonybaloney** [comment]
@jasminSPC this is great work, I'm just reviewing now. In the mean time please can you sign an ICLA and let me know when that's done.http://libcloud.readthedocs.io/en/latest/development.html#contributing-bigger-changes

**Comment by jasmingacic** [comment]
@tonybaloney I've sent signed ICLA to secretary@apache.org today, so I guess I'll have to wait.

File: libcloud/compute/drivers/oneandone.py [hunk 3]
```diff
@@ -0,0 +1,2263 @@
...
+
+        response = self.connection.request(
+            action="servers",
+            data=body,
+            method='POST',
+        )
+
+        return self._to_node(response.object)
+
+    def list_nodes(self):
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py:287]: most drivers have a parameter to filter by location, if the API supports that.

File: libcloud/compute/drivers/oneandone.py [hunk 4]
```diff
@@ -0,0 +1,2263 @@
...
+                    ex_monitoring_policy_id=None,
+                    ex_firewall_policy_id=None,
+                    ex_loadbalancer_id=None,
+                    ex_description=None,
+                    ex_power_on=None):
+        """
+        Creates a node.
+
+        :param name: The name of the new node
+        :type ```str```
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: this should be `:type name: ``str```
you've got too many backticks, also the name of the argument should come after the word type

File: libcloud/compute/drivers/oneandone.py [hunk 5]
```diff
@@ -0,0 +1,2263 @@
...
+
+        :param image: Image from list_images
+        :rtype: ``NodeImage``
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: this should be `:type image:`

File: libcloud/compute/drivers/oneandone.py [hunk 6]
```diff
@@ -0,0 +1,2263 @@
...
+
+        :param ex_loadbalancer_id:
+        :type ```str```
+
+        :param ex_description:
+        :type ```str```
+
+        :param ex_power_on:
+        :type ```bool```
+
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: most of the docstrings are incorrect, please see http://libcloud.readthedocs.io/en/latest/development.html#docstring-conventions

**Review by tonybaloney — CHANGES_REQUESTED** [review]

File: libcloud/compute/drivers/oneandone.py [hunk 7]
```diff
@@ -0,0 +1,2263 @@
...
+        )
+        return response.object
+
+    """
+    Hardware operations
+    """
+
+    def ex_modify_server_hardware(self, server_id,
+                                  fixed_instance_size_id=None, vcore=None,
+                                  cores_per_processor=None, ram=None, ):
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: trailing comma

File: libcloud/compute/drivers/oneandone.py [hunk 8]
```diff
@@ -0,0 +1,2263 @@
...
+
+
+        :return:    Instance of class ``Node``
+        :rtype:     :class: `Node`
+        """
+
+        body = {}
+
+        if size is not None:
+            body[size] = size
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: did you mean  "body['size']"?

File: libcloud/compute/drivers/oneandone.py [hunk 9]
```diff
@@ -0,0 +1,2263 @@
...
+        :rtype: ``str``
+
+        :param password: New password for the server
+
+        :return:    Instance of class ``Node``
+        :rtype:     :class: `Node`
+        """
+
+        body = {
+            "id": image_id,
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: try and stick to single quotes for consistency (and below)

File: libcloud/compute/drivers/oneandone.py [hunk 10]
```diff
@@ -0,0 +1,2263 @@
...
+
+        :param ip_type: Type of the IP address [IPV4,IPV6]
+        :rtype: ``str``
+
+        :return: ``Node`` instance
+        :rtype: ``Node``
+        """
+
+        body = {
+            "type": ip_type
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: single quotes please

File: libcloud/compute/drivers/oneandone.py [hunk 11]
```diff
@@ -0,0 +1,2263 @@
...
+        :rtype: ``str``
+
+        :param firewall_id: ID of the firewall policy
+        :rtype: ``str``
+
+        :return: ``Node`` instance
+        :rtype: ``Node``
+        """
+        body = {
+            "id": firewall_id
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: single quotes

File: libcloud/compute/drivers/oneandone.py [hunk 12]
```diff
@@ -0,0 +1,2263 @@
...
+    Firewall Policy operations
+    """
+
+    def ex_create_firewall_policy(self, name, rules, description=None):
+        """
+        Creates a firewall Policy
+        :param name:
+        :param description:
+        :param rules:
+        :rtype 'dict':  [
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py:812]: I don't think you can do that with sphinx. You can add a nested code block, but the rtype is ``dict``

File: libcloud/compute/drivers/oneandone.py [hunk 13]
```diff
@@ -0,0 +1,2263 @@
...
+        response = self.connection.request(
+            action='shared_storages',
+            method='GET'
+        )
+
+        return response.object
+
+    def ex_get_shared_storage(self, storage_id):
+        """
+                Gets a shared storage
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: please fix nesting

File: libcloud/compute/drivers/oneandone.py [hunk 14]
```diff
@@ -0,0 +1,2263 @@
...
+              "port":"80",
+              "alert_if":"RESPONDING",
+              "email_notification":true
+           }
+        ]
+        :return: Instance of a monitoring policy
+        :rtype: ``dict``
+        """
+
+        body = {"ports": ports}
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: single quotes please

File: libcloud/compute/drivers/oneandone.py [hunk 15]
```diff
@@ -0,0 +1,2263 @@
...
+    def ex_add_monitoring_policy_processes(self, policy_id, processes):
+        """
+        Add monitoring policy processes
+
+        :param policy_id: Id of the monitoring policy
+        :rtype: ``str``
+
+        :param processes: List of processes
+        :rtype: ``list`` of ``dict``
+        [
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py:2059]: as previous

File: libcloud/compute/drivers/oneandone.py [hunk 16]
```diff
@@ -0,0 +1,2263 @@
...
+            driver=self.connection.driver
+        )
+
+    def _to_nodes(self, servers):
+        return [self._to_node(
+            server) for server in servers]
+
+    def _to_node(self, server):
+        extra = {}
+        extra["datacenter"] = server['datacenter']
```
Comment (tonybaloney) [inline libcloud/compute/drivers/oneandone.py]: stick to single quotes in this method

**Review by tonybaloney — CHANGES_REQUESTED** [review]
mostly small style requirements. The docstrings need checking for each method. if you render the docs you'll see what I mean.

**Comment by tonybaloney** [comment]
I also recommend you run `tox -e lint`, TravisCI won't pass this until that stage is passing and I can see some PEP8 problems that would fail it.

File: libcloud/compute/drivers/oneandone.py (hunk 4 above)
Comment (jasmingacic) [reply libcloud/compute/drivers/oneandone.py]: Two ticks instead of three, is that correct?

**Comment by tonybaloney** [comment]
@jasminSPC this has been done to a great standard, well done. Just some small things to fix up then we can get this merged in

File: libcloud/compute/drivers/oneandone.py (hunk 3 above)
Comment (jasmingacic) [reply libcloud/compute/drivers/oneandone.py:287]: The API doesn't support filtering by locations [see here](https://cloudpanel-api.1and1.com/documentation/v1/en/documentation.html#servers_get)

File: libcloud/compute/drivers/oneandone.py (hunk 4 above)
Comment (tonybaloney) [reply libcloud/compute/drivers/oneandone.py]: like this - https://github.com/apache/libcloud/blob/trunk/libcloud/compute/drivers/dimensiondata.py#L894-L916

File: libcloud/compute/drivers/oneandone.py (hunk 12 above)
Comment (jasmingacic) [reply libcloud/compute/drivers/oneandone.py:812]: I'm not sure what do you mean by this. I wanted to give an example of what dict needed to be passed to the function. I can easily remove it.

Comment (tonybaloney) [reply libcloud/compute/drivers/oneandone.py:812]: rtype should just be ``dict`` then in the description of the method you should include a code snippet using the triple >>>
https://github.com/tonybaloney/libcloud/blob/trunk/libcloud/compute/drivers/dimensiondata.py#L277-L288

**Comment by jasmingacic** [comment]
i've replaced triple ticks with double ticks and i replaced double quotes with single quote. 

p.s. My ICLA has been filed.

**Comment by tonybaloney** [comment]
if you can resolve the build errors https://travis-ci.org/apache/libcloud/builds/220768604 we can get this merged.

**Comment by jasmingacic** [comment]
@tonybaloney There you go.

**Comment by tonybaloney** [comment]
👍  LGTM

**Comment by tonybaloney** [comment]
again, @jasminSPC thanks for the contribution. This will take a few minutes to show in the documentation.

**Comment by jasmingacic** [comment]
@tonybaloney WOW this was quick :)

Is there anything else for me to do here?

**Comment by tonybaloney** [comment]
@jasminSPC more contributions are always welcome! A DNS driver?

**Comment by jasmingacic** [comment]
@tonybaloney unfortunately 1&1 doesn't have that feature supported.

**Pull request merged**